### ISSUE TITLE
Organizing files into sub-packages

### ISSUE THREAD

**Issue by brendensoares** [revel/revel#614]
@revel/core What do you think about organizing the go files in the root directory into sub-packages?

I tried simply moving them into sub-folders, but in order for them to share the `revel` package they have to be in the same folder (this is odd and slightly irritating).

We should be able to do this without breaking compatibility and without introducing circular dependencies. Doing this should increase our separation of concerns and hopefully improve code quality through clear lines of division and unit testing.

I also feel that reorganizing into sub-packages will make contribution to Revel more inviting to the Go community and/or new comers.

**Comment by pedromorgan** [comment]
Where are we now with this..

I think there are inbuilt packages/modules eg jobs, testrunner
and then third party

?
its a config issue ??
http://revel-docs.daffodil.uk.com/manual/modules.html

and a manual for others..  

maybe we need `revel create mod` to do that ?

**Comment by brendensoares** [comment]
@pedromorgan see the [Roadmap for v0.12](https://github.com/revel/revel/wiki/Roadmap#v012-see-issues) and specifically the "Create sub-packages" checkbox.

I would like to create a few more internal packages, but @anonx brings up some good points that I haven't thought through fully.

@anonx my high level goal is to have a clear separation of concerns. I'd like to do this without affecting the exported interface that Revel devers see. I want it to be a purely internal change. I will admit that sub-packages may not be the correct solution for each of the items listed above and I am ready to find more reasonable solutions if we can think of them.

As far as criteria for creating a sub-package, I'd say if a segment of code is independently functional and complex enough (say ~300 lines of code or more) then we should seriously consider its re-organization. The criteria is somewhat subjective though.

PS - I am inclined to split v0.12 into 2 releases as it's too large currently. It's more important that we keep consistent momentum than trying to release large improvements. Agreed?

**Comment by brendensoares** [comment]
@anonx it just struck me that we may not be understanding sub-packages the same way. Not all that I've suggested on the v0.12 Roadmap should be external packages. Internal, independently testable packages are more appropriate than external for most, if not all, of the remaining items. Agreed?

**Comment by anonx** [comment]
@brendensoares I'm asking about the criteria just for practical reasons. We have ``init.go`` with the following content:
```go
revel.Filters = []revel.Filter{
	...
	revel.RouterFilter,
	revel.SessionFilter,
	...
}
```
I, as a user of the framework, expect that every standard filter will have predictable import path. For example:
```go
import (
	"github.com/revel/revel/routing"
	"github.com/revel/revel/session"
)
```
So the filters above can be used as:
```go
	...
	routing.Filter,
	session.Filter,
	...
```

If import paths are different I feel like something is wrong with consistency:
```go
revel.Filters = []revel.Filter{
	revel.PanicFilter,
	routing.RouterFilter,
	revel.FilterConfiguringFilter,
	params.Filter,
}
```

> I am inclined to split v0.12 into 2 releases as it's too large currently. It's more important that we keep consistent momentum than trying to release large improvements. Agreed?

:+1:

> Internal, independently testable packages are more appropriate than external for most, if not all, of the remaining items. Agreed?

Yes.

**Comment by brendensoares** [comment]
@anonx to your last point, the value I see is that it provides a place to put community driven modules and samples and alleviates the clutter in the core library. It does aggravate the package management issue aka version dependency, but that's a good thing; we need to decide a solution for that ASAP.

In regards to splitting the release, any good ideas on a release theme to put on the Roadmap?

Lastly, I didn't expect a dever to have to import any of the internal sub-packages, but maybe that was an oversight on my part. The `session` package has more than just a filter; it has/needs an interface to allow other backend drivers i.e. Redis, Files, In-memory etc. We could put the `revel.SessionFilter` in a `filters` sub-package and simply have it import a subsequent `session` package that contains the interface and other backend drivers. That way the `revel` package still owns the filter interface implementations and we can further modularize independent code i.e. sessions, routing, flash panic/error handling. Agreed?

Either way, it would be productive if we could chat directly for 30 minutes to outline RFCs for each of the remaining code re-organization efforts aka sub-packages.

**Comment by brendensoares** [comment]
@NitroKKX I've chatted with @anonx and we decided it's best to keep `watcher` in the `revel` repo to ensure `watcher` functionality when not using the `revel` command.

Here is the [outcome of our chat](https://github.com/revel/revel/wiki/Meetings-2015-01-03) as far as new sub-packages:
- sub-package naming convention vs Revel API exported attributes
  - `testing`
  - `binding`
    - decided: combine binder and params into 1 package
  - `routing`
    - decided: move router.go
  - decided: no `server`
  - `validation`
    - decided: go ahead and move it to subpack
    - decided: Validation to validation.Context
  - `watcher`
    - decided: Watcher to watcher.Notifier
  - `session`
    - decided: session.Session to session.Values
    - decided: SessionFilter to session.Filter
    - merge flash.go to session i.e. FlashFitler with session.Filter
  - `controller`
    - decided: Controller to controller.Context and controller.NewContext
    - decided: pass in config.Context instance to NewController
  - `filter`
    - decided: remove initial var Filters values and require app/init.go to set them or panic
    - decided: Filter to filter.Type
    - decided: Filters to filter.Stack
  - `config` for app config...needs eventual abstraction for backend drivers
    - decided: MergedConfig to config.Context
    - decided: NewEmptyConfig to config.NewContext
    - decided: LoadConfig to config.LoadContext
    - need a revel/cmd issue
  - should make sub-packages more independent of Revel in the future

**Comment by NitroKKX** [comment]
Makes sense to me. Thanks

**Comment by brendensoares** [comment]
After trying to combine `params` and `binder` into `binding` I've decided they are separate concerns and should be packaged separately.

`binder`'s input is a string and its output is a Go variable. I'm wondering if `binder` is the most appropriate name for this. It seems `pack` and `unpack` makes more sense than `bind` and `unbind`. It's main purpose is the conversion of strings to and from variables, right? In my experience, binding has to do with event processing more than data conversion. Thoughts?

So, it seems we'll need a `params` sub-package after all @anonx. It's input is a `net/http` `Request` and its output is a key/value list of variables. I can't think of a better name for the package or any sort of `*ing` equivalent. I'm open to ideas. Same goes for `binder`, but I don't mind keeping `binding`.

**Comment by notzippy** [comment]
Rather than binding why not use encoding ? That is the standard used in golang for json and xml translations from strings to golang objects. It should be applicable here as well.

**Comment by brendensoares** [comment]
@notzippy that makes sense. What about `params`?

Any input @pushrax @anonx ?

**Comment by anonx** [comment]
There are `revel.TRACE`s, `revel.INFO`s, `revel.WARN`s, and `revel.ERROR`s all around the code. Looks like we cannot go without a separate `log` package.

**Comment by notzippy** [comment]
For logs, we could define it as an interface in revel and implement it in a different package. From what I have seen with logging there are many different approaches to take, rather than attempt to have revel do it all why not allow someone to implement an interface to there favorite logging package and assign it in the config. By default revel would use the golang logger, (which would be in its own package)

**Comment by brendensoares** [comment]
@anonx we can consider the use of dependency injection for challenges like this, but in this case, I don't think we even need the logger. It seems the better solution is to return errors where applicable instead of logging the errors directly. Revel can do what it wants with the errors. Agreed?

@notzippy the new logger interface is planned for [v0.15](https://github.com/revel/revel/wiki/Roadmap#v015-see-issues). It's a need, but not as important as other items.

**Comment by anonx** [comment]
@brendensoares Yes, good idea. But sounds like a lot of work.

**Comment by brendensoares** [comment]
@anonx I'm almost ready to push the PR for review. I have one last `Params` dependency in `binding` to solve and then I can show you my progress.

**Comment by anonx** [comment]
@brendensoares Looking forward to it.

**Comment by brendensoares** [comment]
#871

**Comment by jeevatkm** [comment]
I agree with organizing, if possible making independent as possible.

But my question is about naming and structure, for example:
```go
import (
    "github.com/revel/revel/routing"
    "github.com/revel/revel/session"
)
``` 

Instead, it's better have structure/name like this:
```go
import (
    "github.com/revel/log"
    "github.com/revel/routing"
    "github.com/revel/session"
)
```
